services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/UploadWithKnownContentLengthHelper.java

```diff
@@ -44,10 +53,10 @@
 public final class UploadWithKnownContentLengthHelper {
     private static final Logger log = Logger.loggerFor(UploadWithKnownContentLengthHelper.class);
 
+    private static final ConcurrentHashMap<Integer, CompletedPart> EXISTING_PARTS = new ConcurrentHashMap<>();
```

Comment:
I've been looking for tests that test the different code paths in this class, but there is no test class for this class. There are TM functional tests, but I think it would be valuable to have direct unit tests of this class now that it's pretty complex - not sure why they were not included from the beginning. We can discuss offline.

Reply:
this class is covered by `UploadObjectHelperTest`, added a test there and also created a test class for `KnownContentLengthAsyncRequestBodySubscriber`